### ISSUE TITLE
Cannot add Permission Override for own role

### ISSUE THREAD

**Issue by Ryeera** [DV8FromTheWorld/JDA#1523]
## General Troubleshooting
  
- [✔️] I have checked for similar issues.
- [✔️] I have updated to the [latest JDA version][download].
- [✔️] I have checked the branches or the maintainers' PRs for upcoming bug fixes.

## Bug Report

I wrote a bot that is supposed to make a Text-Channel for each Voice-Channel, that is hidden by default and becomes visible when a member joins the Voice-Channel. Since I don't want to make my bot public while requiring Admin-Permissions, I made it so in a `ChannelAction`, the bot creates the new Text-Channel and adds a permission-override for his own role so that it can always see the channel. Afterwards, I made it so it removes the `View Channel` permission from `@everyone` so that only the bot (well and admins) can see that channel.

I gave the Bot the `Manage Roles` permission, which, according to my Discord-Client "allows members to change permissions of individual channels that they have access to." And as expected, when I create a Text-Channel with a Client for which I also made a role with the exact same permissions as my bot, and make it private with only that role having access, it works perfectly fine and I can make permission-overrides.

Now to the actual bug (I think?): When I make a `ChannelAction<TextChannel>` and in there make a permission-override for `@everyone`, it doesn't work. I get the response: 
`net.dv8tion.jda.api.exceptions.InsufficientPermissionException: You must have Permission.MANAGE_PERMISSIONS on the channel explicitly in order to set permissions you don't already have!`

### Expected Behavior

I expected my bot to create a new channel that only its role has access to.

### Code Example or Reproduction Steps

My code:
```java
EnumSet<Permission> readPerms = EnumSet.of(Permission.MESSAGE_READ);
ChannelAction<TextChannel> action = guild.createTextChannel("test");
    action.addPermissionOverride(guild.getBotRole(), readPerms, null);
    action.queue(tc -> {
        tc.putPermissionOverride(guild.getPublicRole()).deny(readPerms).queue();
});
```

What I want to do, reproduced with a cl:
1. Make a Role for a client with the same permissions as the bot.
2. Give the role to a client without any other roles.
3. With that client, create a new Text-Channel
4. Enter a name for that channel
5. Flip the switch for Private Channel
6. Select the role that was given to the client (note how you can't select the client directly)
7. Click on "Create Channel"
8. Right-click on the new channel and select "Edit Channel"
9. Go to permissions
10. Open Advanced Permissions
11. You can now, without a problem, create an override for a single member.

### Code for JDABuilder or DefaultShardManagerBuilder Used

```java
JDABuilder builder = JDABuilder.create(GatewayIntent.GUILD_MESSAGES, GatewayIntent.GUILD_VOICE_STATES, GatewayIntent.GUILD_MEMBERS);
builder.enableCache(CacheFlag.MEMBER_OVERRIDES);
builder.disableCache(CacheFlag.ACTIVITY, CacheFlag.EMOTE, CacheFlag.CLIENT_STATUS);
builder.setToken(config.getString("token"));
builder.setActivity(Activity.watching("the VoiceChannels"));
builder.setMemberCachePolicy(MemberCachePolicy.ALL);
builder.addEventListeners(new DragoBot());
try {
    jda = builder.build();
    jda.awaitReady();
} catch (LoginException | InterruptedException e) {
    logger.logStackTrace(e);
    System.exit(1);
}
```

### Exception or Error

```
[JDA MainWS-ReadThread] ERROR JDA - One of the EventListeners had an uncaught exception
net.dv8tion.jda.api.exceptions.InsufficientPermissionException: You must have Permission.MANAGE_PERMISSIONS on the channel explicitly in order to set permissions you don't already have!
        at net.dv8tion.jda.internal.requests.restaction.ChannelActionImpl.addOverride(ChannelActionImpl.java:286)
        at net.dv8tion.jda.internal.requests.restaction.ChannelActionImpl.addRolePermissionOverride(ChannelActionImpl.java:199)
        at net.dv8tion.jda.internal.requests.restaction.ChannelActionImpl.addRolePermissionOverride(ChannelActionImpl.java:40)
        at net.dv8tion.jda.api.requests.restaction.ChannelAction.addPermissionOverride(ChannelAction.java:299)
        at net.dv8tion.jda.api.requests.restaction.ChannelAction.addPermissionOverride(ChannelAction.java:251)
        at de.Ryeera.DragoBot.DragoBot.initialize(DragoBot.java:344)
        at de.Ryeera.DragoBot.DragoBot.onGuildMessageReceived(DragoBot.java:545)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at net.dv8tion.jda.api.hooks.ListenerAdapter.onEvent(ListenerAdapter.java:405)
        at net.dv8tion.jda.api.hooks.InterfacedEventManager.handle(InterfacedEventManager.java:96)
        at net.dv8tion.jda.internal.hooks.EventManagerProxy.handleInternally(EventManagerProxy.java:88)
        at net.dv8tion.jda.internal.hooks.EventManagerProxy.handle(EventManagerProxy.java:70)
        at net.dv8tion.jda.internal.JDAImpl.handleEvent(JDAImpl.java:149)
        at net.dv8tion.jda.internal.handle.MessageCreateHandler.handleInternally(MessageCreateHandler.java:97)
        at net.dv8tion.jda.internal.handle.SocketHandler.handle(SocketHandler.java:36)
        at net.dv8tion.jda.internal.requests.WebSocketClient.onDispatch(WebSocketClient.java:952)
        at net.dv8tion.jda.internal.requests.WebSocketClient.onEvent(WebSocketClient.java:839)
        at net.dv8tion.jda.internal.requests.WebSocketClient.handleEvent(WebSocketClient.java:817)
        at net.dv8tion.jda.internal.requests.WebSocketClient.onBinaryMessage(WebSocketClient.java:990)
        at com.neovisionaries.ws.client.ListenerManager.callOnBinaryMessage(ListenerManager.java:385)
        at com.neovisionaries.ws.client.ReadingThread.callOnBinaryMessage(ReadingThread.java:276)
        at com.neovisionaries.ws.client.ReadingThread.handleBinaryFrame(ReadingThread.java:996)
        at com.neovisionaries.ws.client.ReadingThread.handleFrame(ReadingThread.java:755)
        at com.neovisionaries.ws.client.ReadingThread.main(ReadingThread.java:108)
        at com.neovisionaries.ws.client.ReadingThread.runMain(ReadingThread.java:64)
        at com.neovisionaries.ws.client.WebSocketThread.run(WebSocketThread.java:45)
```

**Comment by Andre601** [comment]
The error points out clearly, what is wrong.
Your bot needs the `Manage Permissions` permission in that particular category, to be able to create channels and manage those.

This is a recent change in Discord's API. Bots now require the `Manage Permission` right alongside any other necessary permission to alter channel-specific ones.

**Comment by MinnDevelopment** [comment]
You should do this instead:

```java
EnumSet<Permission> readPerms = EnumSet.of(Permission.MESSAGE_READ);
guild.createTextChannel("test")
  .addPermissionOverride(guild.getSelfMember(), readPerms, null)
  .addPermissionOverride(guild.getPublicRole(), null, readPerms)
  .queue();
```

**Comment by Ryeera** [comment]
@MinnDevelopment That's exactly what I did before, what I did in the example I provided was one my attempts to force the own read-permission to be applied first.

@Andre601 so what you're saying is, is that I need to give my bot Admin-permissions? I could also tell people who add my bot to give the channels that they want to be seen by the bot the appropriate permissions, but that's not very user-friendly.

**Comment by Andre601** [comment]
> @Andre601 so what you're saying is, is that I need to give my bot Admin-permissions? I could also tell people who add my bot to give the channels that they want to be seen by the bot the appropriate permissions, but that's not very user-friendly.
> I didn't know that bots would use permissions differently from clients, that is so weird and makes testing and confirming things hard...

No. All you need to do is give the bot `Manage Permission` rights for the category, where you want to create and update channels.
Admin isn't required for this at all, just the right perms. Unfortunately is the Manage Permission right not a global one from what I saw, so you can't request it through the bot invite, but it may work, if your bot has sufficient perms (I think Manage Server should work) to create a new category where it sets the rights.

That's at least my bare-bones knowledge about this whole thing

**Comment by MinnDevelopment** [comment]
The error you get in this case indicates that your bot does not have `MESSAGE_READ` on any of its roles in the first place, this means it also cannot set it on any of the channels it creates.

**Comment by Ryeera** [comment]
> > @Andre601 so what you're saying is, is that I need to give my bot Admin-permissions? I could also tell people who add my bot to give the channels that they want to be seen by the bot the appropriate permissions, but that's not very user-friendly.
> > I didn't know that bots would use permissions differently from clients, that is so weird and makes testing and confirming things hard...
> 
> No. All you need to do is give the bot `Manage Permission` rights for the category, where you want to create and update channels.
> Admin isn't required for this at all, just the right perms. Unfortunately is the Manage Permission right not a global one from what I saw, so you can't request it through the bot invite, but it may work, if your bot has sufficient perms (I think Manage Server should work) to create a new category where it sets the rights.


My bot is not supposed to move all the VCs into a separate category and it's also not supposed to make new voice-channels, I just want it to make a Text-Channel for each Voice-Channel (in the same category) and then put a Permission-Override in it that makes it so the bot can see the channel (which according to the error-message I got should work because I *do* have the `Read Messages` as well as `Send Message` permissions and *then* it makes it so `@everyone` can't see the channel (deny `View Channel`). I'm not sure if this is intended behavior by Discord but if it is, I need to give my bot admin to achieve what I want the bot to do. 😢 

Also @MinnDevelopment here's the permissions my bot has in its role:
- Manage Roles
- Manage Channels
- Read Messages
- Send Messages
- Embed Links
- Move Members

**Comment by MinnDevelopment** [comment]
I have setup a server with the same permissions and cannot reproduce this.

**Comment by Ryeera** [comment]
Alright, at least I narrowed down the problem. In public channels it's not a problem, but in channels where the access is limited to one or more roles, it doesn't work, *even when I add the bot's role*! And that last part is what I don't get. Bot has manage permissions, it isn't revoked anywhere in the channel-permissions, and even when the bot has explicit access to the channel, it doesn't work. And most confusingly, even when I give the bot explicit permissions for `Manage Permissions`, it STILL doesn't work! I'm at the end of my knowledge and testing now... If  you don't have any ideas, it's okay, you can close the request, and I'll just publish my bot with admin-perms required.

**Comment by Ryeera** [comment]
Guys I actually figured it out. It is still the weirdest thing tho and I'm not sure if it is intended behaviour. I will post it tomorrow because I need to go into detail and I need to sleep now. Just posting this so the issue isn't closed yet. Gonna close this tomorrow then.

**Comment by Ryeera** [comment]
So basically, this was the solution:

This works always:
```java
ChannelAction<TextChannel> action = guild.createTextChannel(guildConfig.getString("Names").replace("{vc}", vc.getName()));
action.setTopic(guildConfig.getString("Descriptions").replace("{vc}", vc.getName()));
action.addPermissionOverride(guild.getBotRole(), readPerms, null);
action.addPermissionOverride(guild.getPublicRole(), null, readPerms);
if (vc.getParent() != null) {
	action.setParent(vc.getParent());
}
action.queue();
```

This works **only if the Voice-Channel in question is public!** The permissions of the underlying category make NO DIFFERENCE! Even if the category and Voice-Channel have explicit permissions for the bot to `Manage Permissions`, it still gives me the aforementioned error:
```java
ChannelAction<TextChannel> action = guild.createTextChannel(guildConfig.getString("Names").replace("{vc}", vc.getName()));
action.setTopic(guildConfig.getString("Descriptions").replace("{vc}", vc.getName()));
if (vc.getParent() != null) {
	action.setParent(vc.getParent());
}
action.addPermissionOverride(guild.getBotRole(), readPerms, null);
action.addPermissionOverride(guild.getPublicRole(), null, readPerms);
action.queue();
```

So basically, it matters, when I add the permission-override. Before, or after I set the category. I don't know if this is intended behavior or if this even makes any sense. but I finally figured it out and I can continue with making my bot. Thank you for your help everyone!

**Comment by MinnDevelopment** [comment]
This should be resolved with **4.2.0_229**.
